Sources/ApplicativeRouter/SyntaxRouter.swift

```diff
@@ -0,0 +1,159 @@
+import Optics
+
+// MARK: - Syntax Router
+
+// TODO: should this be generic over any monoid `M` instead of using `RequestData` directly?
+// TODO: generic over an error semigroup too?
+public struct Router<A> {
+  let parse: (RequestData) -> (rest: RequestData, match: A)?
+  let print: (A) -> RequestData?
+  let template: (A) -> RequestData?
+
+  public func match(request: URLRequest) -> A? {
+    return (self <% _end).parse(route(from: request))?.match
+  }
+
+  public func match(url: URL) -> A? {
+    return match(request: URLRequest(url: url))
+  }
+
+  public func match(string: String) -> A? {
+    return URL(string: string).flatMap(match(url:))
+  }
+
+  public func request(for a: A) -> URLRequest? {
+    return self.print(a).flatMap(request(from:))
+  }
+
+  public func url(for a: A) -> URL? {
+    return self.print(a).flatMap(request(from:)).flatMap { $0.url }
+  }
+
+  public func templateRequest(for a: A) -> URLRequest? {
+    return self.template(a).flatMap(request(from:))
+  }
+
+  public func templateUrl(for a: A) -> URL? {
+    return self.templateRequest(for: a).flatMap { $0.url }
+  }
+}
+
+// Functor
+
+extension Router {
+  public func map<B>(_ f: PartialIso<A, B>) -> Router<B> {
+    return f <¢> self
+  }
+
+  public static func <¢> <B> (lhs: PartialIso<A, B>, rhs: Router) -> Router<B> {
+    return Router<B>(
+      parse: { route in
+        guard let (rest, match) = rhs.parse(route) else { return nil }
+        return lhs.image(match).map { (rest, $0) }
+      },
+      print: lhs.preimage >-> rhs.print,
+      template: lhs.preimage >-> rhs.template
+    )
+  }
+}
+
+// Apply
+
+extension Router {
+  /// TODO: use `Tuple`?
+  /// Processes with the left and right side routers, and if they succeed returns the pair of their results.
+  public static func <%> <B> (lhs: Router, rhs: Router<B>) -> Router<(A, B)> {
+    return Router<(A, B)>(
+      parse: { str in
+        guard let (more, a) = lhs.parse(str) else { return nil }
+        guard let (rest, b) = rhs.parse(more) else { return nil }
+        return (rest, (a, b))
+      },
+      print: { ab in
+        let lhsPrint = lhs.print(ab.0)
+        let rhsPrint = rhs.print(ab.1)
+        return (curry(<>) <¢> lhsPrint <*> rhsPrint) ?? lhsPrint ?? rhsPrint
+      },
+      template: { ab in
+        let lhsPrint = lhs.template(ab.0)
+        let rhsPrint = rhs.template(ab.1)
+        return (curry(<>) <¢> lhsPrint <*> rhsPrint) ?? lhsPrint ?? rhsPrint
+    })
+  }
+
+  /// Processes with the left and right side routers, discarding the result of the left side.
+  public static func %> (x: Router<Prelude.Unit>, y: Router) -> Router {
+    return (PartialIso.commute >>> PartialIso.unit.inverted) <¢> x <%> y
+  }
+}
+
+extension Router where A == Prelude.Unit {
+  /// Processes with the left and right routers, discarding the result of the right side.
+  public static func <% <B>(x: Router<B>, y: Router) -> Router<B> {
+    return PartialIso.unit.inverted <¢> x <%> y
```

Comment:
the `<%` and `%>` operators are precisely like `<*` and `*>` except they can _only_ work when the data being discarded is `Unit`. that's kinda neat